File: Algo/Leetcode/400NthDigit.cpp
class Solution {
public:
    int findNthDigit(int n) {
        int digit = 1; long long start = 1; long long count = 9;
        while(n > count){
            n -= count;
            start *= 10;
            digit += 1;
            count = start * digit * 9;
        }
        int num = start + (n - 1) / digit;
        return to_string(num)[(n-1)%digit] - '0';
    }
};

File: Algo/Leetcode/400NthDigit.py
class Solution(object):
    def findNthDigit(self, n):
        """
        :type n: int
        :rtype: int
        """
        digit = 0

        while n > 9 * 10 ** digit * (digit + 1):
            n -= 9 * 10 ** digit * (digit + 1)
            digit += 1

        start = 10 ** digit
        remain = start + (n - 1) / (digit + 1)

        return int(str(remain)[(n - 1) % (digit + 1)])
